Do not simplify (T)bv-typed when T has smaller bit width

tautschnig opened the pull request:
We must not modify the bit width of bv (and not signed/unsigned bv) typed expressions via type casts. Make sure simplification doesn't given rise to such expressions.

Fixes: #7426

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Support (T)bv-typed conversion when T has smaller bit width

We can safely reduce the number of bits of a bv (and not just
signed/unsigned bv) typed expression by just cutting off
more-significant bits.

Fixes: #7426

kroening commented:
BTW, why is the simplification wrong?

tautschnig commented:
> BTW, why is the simplification wrong?

https://github.com/diffblue/cbmc/blob/develop/src/solvers/flattening/boolbv_typecast.cpp#L429..L433 claims it is. But then maybe the typecast flattening code is wrong?

kroening commented:
I don't think that the invariant is claiming that the cast would necessarily be unsound.

A question is what semantics to give it. I'd say it's straight forward when casting to something smaller. You drop the extra bits.

There is certainly ambiguity when casting to something bigger, since it's not obvious whether to zero extend or sign extend.

tautschnig committed:
clang-format boolbv_typecast.cpp

The nested `switch` statements had indentation mixed up, making it very
difficult to understand which `switch` statement a `case` actually
belonged to.
